addons/mail/static/src/models/thread/thread.js

```diff
@@ -462,20 +496,25 @@ function factory(dependencies) {
 
         /**
          * Mark the specified conversation as read/seen.
+         *
+         * @param {integer} messageId the message to be considered as last seen, by default
+         * last message of the thread.
          */
-        async markAsSeen() {
-            if (this.message_unread_counter === 0) {
+        async markAsSeen(messageId) {
+            if (this.pendingSeenMessageId && messageId <= this.pendingSeenMessageId) {
                 return;
             }
-            if (this.model === 'mail.channel') {
-                const seen_message_id = await this.async(() => this.env.services.rpc({
-                    model: 'mail.channel',
-                    method: 'channel_seen',
-                    args: [[this.id]]
-                }, { shadow: true }));
-                this.update({ seen_message_id });
+            if (
+                this.lastSeenByCurrentPartnerMessageId &&
+                messageId <= this.lastSeenByCurrentPartnerMessageId
+            ) {
+                return;
             }
-            this.update({ message_unread_counter: 0 });
+            this.update({ pendingSeenMessageId: messageId });
+            if (this.model !== 'mail.channel') {
+                return;
+            }
```

Comment:
strange order here, shouldn't the preventive return be done at the start of the method?